src/grain/ledger.js

```diff
@@ -0,0 +1,220 @@
+} from "./grain";
+
+export opaque type EventOrderT = symbol;
+export const EventOrder: {|
+  +ASCENDING: EventOrderT,
+  +DESCENDING: EventOrderT,
+|} = deepFreeze({
+  ASCENDING: Symbol("ASCENDING"),
+  DESCENDING: Symbol("DESCENDING"),
+});
+
+export opaque type EventSortT = symbol;
+export const EventSort: {|
+  +AMOUNT: EventSortT,
+  +TIMESTAMP: EventSortT,
+|} = deepFreeze({
+  AMOUNT: Symbol("AMOUNT"),
+  TIMESTAMP: Symbol("TIMESTAMP"),
+});
+
+export type EventsOptions = {|
+  // The address to filter for; events will be returned if this address is
+  // involved at all (e.g. as a recipient or sender).
+  // Only exact address matches are included (no prefix matching).
+  // If null, all addresses are included.
+  +address: NodeAddressT | null,
+  // The order to return results in (ASCENDING or DESCENDING).
+  // Defaults to ASCENDING.
+  +eventOrder: EventOrderT,
+  // The sorting strategy to use (by TIMESTAMP or by AMOUNT).
+  // Defaults to TIMESTAMP.
+  +eventSort: EventSortT,
+|};
+
+export const DEFAULT_EVENTS_OPTIONS = deepFreeze({
+  address: null,
+  eventOrder: EventOrder.ASCENDING,
+  eventSort: EventSort.TIMESTAMP,
+});
+
+/**
+ * Models a transfer of grain from a sender to recipient.
+ * A grainholder may transfer grain to themself (which is a no-op).
+ */
+export type Transfer = {|
+  +type: "TRANSFER",
+  +version: string,
+  +sender: NodeAddressT,
+  +recipient: NodeAddressT,
+  +amount: Grain,
+  +timestampMs: number,
+|};
+
+/**
+ * Models a participant recieving newly minted grain.
+ */
+export type Distribution = {|
+  +type: "DISTRIBUTION",
+  +version: string,
+  +recipient: NodeAddressT,
+  +amount: Grain,
+  +timestampMs: number,
+|};
+
+export type LedgerEvent = Distribution | Transfer;
+
+/**
+ * A ledger tracks balances, earnings, and the event history of participants.
+ *
+ * We may refactor the `events` to be an async generator in the future,
+ * when we stop storing all the events in memory.
+ */
+export interface Ledger {
+  /**
+   * Stores the current grain balance of each address.
+   */
+  balances(): Map<NodeAddressT, Grain>;
+  /**
+   * Stores the lifetime earnings of each address.
+   *
+   * Necessary for computing which participants have been underpaid.
+   */
+  earnings(): Map<NodeAddressT, Grain>;
+  /**
+   * Retrieve the events.
+   *
+   * Can be filtered and sorted based on the
+   * EventsOptions, see that type for documentation.
+   */
+  events(?EventsOptions): $ReadOnlyArray<LedgerEvent>;
+}
+
+export class InMemoryLedger implements Ledger {
+  _balances: Map<NodeAddressT, Grain>;
+  _earnings: Map<NodeAddressT, Grain>;
+  _events: $ReadOnlyArray<LedgerEvent>;
+
+  /**
+   * Construct a Ledger from an array of in-memory events.
+   *
+   * The events must be in timestamp sorted order.
+   */
+  constructor(events: $ReadOnlyArray<LedgerEvent>) {
+    this._balances = new Map();
+    this._earnings = new Map();
+    this._events = events;
+    let lastTimestamp = -Infinity;
```

Comment:
When are we updating this timestamp?

Reply:
Actually, if this turns out to be a bug, I think that would be an argument for
https://eslint.org/docs/rules/prefer-const#suggest-using-const-prefer-const

Reply:
Yeah, it was a bug. Fixed, and I think that lint rule is a fine idea.